src/set_theory/surreal.lean

```diff
@@ -286,12 +454,219 @@ instance : ordered_add_comm_group surreal :=
+end
+
+@[simp] theorem pow_log_eq_self {a : ℤ} (n : @submonoid.powers ℤ _ a) : pow a (log n) = n :=
+begin
+  unfold pow,
+  unfold log,
+  rcases n with ⟨_ , hn⟩,
+  congr,
+  exact classical.some_spec hn,
+end
+
+lemma nsmul_pow_two_pow_half (n : ℕ) : 2 ^ n • pow_half n = 1 :=
+begin
+  induction n with n hn,
+  { simp only [nsmul_one, pow_half_zero, nat.cast_one, pow_zero] },
+  { rw [← hn, ← double_pow_half_succ_eq_pow_half n, smul_smul (2^n) 2 (pow_half n.succ),
+        mul_comm, pow_succ] }
+end
+
+lemma nsmul_pow_two_pow_half' (n k : ℕ) : 2 ^ n • pow_half (n + k) = pow_half k :=
+begin
+  induction k with k hk,
+  { simp [nsmul_pow_two_pow_half] },
+  { rw [← double_pow_half_succ_eq_pow_half (n + k), ← double_pow_half_succ_eq_pow_half k] at hk,
+    have : 2 ^ n • 2 • pow_half (n + k).succ = 2 • 2 ^ n • pow_half (n + k).succ,
+      by { apply smul_algebra_smul_comm },
+    rw this at hk,
+    exact (@gmul_cancel_of_pos 2 (2^n • pow_half (n + k).succ) (pow_half k.succ) (by norm_num)) hk }
+end
+
+lemma nsmul_int_pow_two_pow_half (m : ℤ) (n k : ℕ) :
+  (m * 2 ^ n) • pow_half (n + k) = m • pow_half k :=
+begin
+  rw mul_gsmul,
+  congr,
+  norm_cast,
+  exact nsmul_pow_two_pow_half' n k,
+end
+
+lemma dyadic_aux {m₁ m₂ : ℤ} {y₁ y₂ : ℕ} (h₂ : m₁ * (2 ^ y₁) = m₂ * (2 ^ y₂)) :
+  m₁ • pow_half y₂ = m₂ • pow_half y₁ :=
+begin
+  by_cases y₁ ≤ y₂,
+  { obtain ⟨c, hc⟩ := le_iff_exists_add.1 h,
+    rw [hc, add_comm, pow_add, ← mul_assoc, mul_eq_mul_right_iff] at h₂,
+    cases h₂,
+    { rw [h₂, hc, add_comm, nsmul_int_pow_two_pow_half m₂ c y₁] },
+    { have := nat.one_le_pow y₁ 2 nat.succ_pos',
+      linarith } },
+  { push_neg at h,
+    obtain ⟨c, hc⟩ := le_iff_exists_add.1 (le_of_lt h),
+    rw [hc, add_comm, pow_add, ← mul_assoc, mul_eq_mul_right_iff] at h₂,
+    cases h₂,
+    { rw [← h₂, hc, add_comm, nsmul_int_pow_two_pow_half m₁ c y₂] },
+    { have := nat.one_le_pow y₂ 2 nat.succ_pos',
+      linarith } }
+end
+
+/-- The map `dyadic_map` sends [m, 2^n] to m • half ^ n. -/
+noncomputable def dyadic_map : localization (@submonoid.powers ℤ _ 2) → surreal :=
+begin
+  apply quotient.lift,
+  swap,
+  { rintro ⟨m, n⟩,
+    exact m • pow_half (log n) },
+  { rintros ⟨m₁, n₁⟩ ⟨m₂, n₂⟩ h₁,
+    obtain ⟨⟨n₃, y₃, hn₃⟩, h₂⟩ := localization.r_iff_exists.1 h₁,
+    simp only [subtype.coe_mk, mul_eq_mul_right_iff] at h₂,
+    cases h₂,
+    { simp only,
+      obtain ⟨a₁, ha₁⟩ := classical.indefinite_description _ n₁.prop,
+      obtain ⟨a₂, ha₂⟩ := classical.indefinite_description _ n₂.prop,
+      have hn₁ : n₁ = pow 2 a₁, by { ext1, solve_by_elim },
+      have hn₂ : n₂ = pow 2 a₂, by { ext1, solve_by_elim },
+      rw [hn₁, hn₂, log_pow_eq_self 2 rfl.ge, log_pow_eq_self 2 rfl.ge],
+      apply dyadic_aux,
+      rwa [ha₁, ha₂] },
+    { have := nat.one_le_pow y₃ 2 nat.succ_pos',
+      linarith } }
+end
+
+/-- We define dyadic surreals as the image of the map `dyadic_map`. -/
+def dyadic := set.image dyadic_map
```

Comment:
This has type `set (localization (@submonoid.powers ℤ _ 2)) → set surreal` - is this really what you meant?

Whether it's what you meant or not, can you add an explicit type here to make it clear what you actually meant?
```suggestion
/-- We define dyadic surreals as the image of the map `dyadic_map`. -/
def dyadic : sorry := set.image dyadic_map
```

Reply:
Thanks! I'll add the type. I simply want to define `dyadic` as a subtype of `surreal` which mathematically is the image of the map `dyadic_map`. 

I'll prove next that `dyadic_map` is injective and gives an embedding. I think then this should be replaced by `function.embedding.image`.